harness/tests/cli/test_auth.py

```diff
@@ -43,9 +43,263 @@ def test_auth_with_store(requests_mock: requests_mock.Mocker, user: Optional[str
+        ScenarioSet("n", "y", "*", "*", "*", "y", Check("cache")),
+    ],
+)
+@mock.patch("determined.common.api.authentication._is_token_valid")
+@mock.patch("determined.common.api.authentication.do_login")
+@mock.patch("getpass.getpass")
+@mock.patch("determined.common.api.authentication.TokenStore.get_active_user")
+@mock.patch("determined.common.api.authentication.TokenStore.get_token")
+@mock.patch("determined.common.api.authentication.TokenStore.drop_user")
+@mock.patch("determined.common.api.authentication.TokenStore.set_token")
+def test_login_scenarios(
+    # Don't care about set_token and it doesn't return anything.
+    _: mock.MagicMock,
+    # Don't care about drop_user and it doesn't return anything.
+    __: mock.MagicMock,
+    mock_get_token: mock.MagicMock,
+    mock_get_active_user: mock.MagicMock,
+    mock_getpass: mock.MagicMock,
+    mock_do_login: mock.MagicMock,
+    mock_is_token_valid: mock.MagicMock,
+    scenario_set: ScenarioSet,
+) -> None:
+    def getpass(*_: Any) -> str:
+        return "prompt_pass"
+
+    def _is_token_valid(master_url: str, token: str, cert: Any) -> bool:
+        return token in ["cache", "env_token"]
+
+    def do_login(*_: Any) -> str:
+        return "new_token"
+
+    mock_getpass.side_effect = getpass
+    mock_is_token_valid.side_effect = _is_token_valid
+    mock_do_login.side_effect = do_login
+
+    for ru, rp, eu, ep, et, cache in scenario_set.scenarios():
+        # Convert short-codes from case definitions into real values.
+        req_user = {"y": "user", "n": None}[ru]
+        req_pass = {"y": "req_pass", "n": None}[rp]
+        env_user = {"y": "user", "e": "extra", "n": None}[eu]
+        env_pass = {"y": "env_pass", "n": None}[ep]
+        env_token = {"y": "env_token", "n": None}[et]
+
+        def get_active_user(*_: Any) -> Optional[str]:
+            return None if cache == "n" else "user"  # noqa:B023
+
+        def get_token(*_: Any) -> Optional[str]:
+            if cache == "n":  # noqa:B023
+                return None
+            if cache == "x":  # noqa:B023
+                return "expired"
+            if cache == "y":  # noqa:B023
+                return "cache"
+            raise ValueError(f"unexpected cache value {cache!r}")  # noqa:B023
+
+        mock_get_active_user.side_effect = get_active_user
+        mock_get_token.side_effect = get_token
+
+        mock_is_token_valid.reset_mock()
+        mock_do_login.reset_mock()
+
+        old_det_user = os.environ.get("DET_USER")
+        old_det_pass = os.environ.get("DET_PASS")
+        old_det_token = os.environ.get("DET_USER_TOKEN")
+        try:
+            setenv_optional("DET_USER", env_user)
+            setenv_optional("DET_PASS", env_pass)
+            setenv_optional("DET_USER_TOKEN", env_token)
+
+            auth = authentication.Authentication("master_url", req_user, req_pass, None)
+
+            # Make sure we got the results we expected.
```

Comment:
I wonder if it would be useful to add a check for the exact `UsernameTokenPair` that gets returned from the authentication logic, at least for some of the test cases. All of these tests just check that we return the correct token, but it could be super useful to be sure that we also returned the correct username.